quantum/impl/quantum_read_write_mutex_impl.h

```diff
@@ -180,24 +217,54 @@ inline
 bool ReadWriteMutex::ReadGuard::tryLock()
 {
     
-    if (_ownsLock) return true;
-    assert(_mutex);
+    assert(_mutex && !ownsLock());
     _ownsLock = _mutex->tryLockRead();
     return _ownsLock;
 }
 
+inline
+void ReadWriteMutex::ReadGuard::upgradeToWrite()
```

Comment:
With the upgradeToWrite feature of ReadGuard, we don't seem to need WriteGuard because it can be achieved via
ReadGuard g(...);
g.upgradeToWrite();
Is that so? 

Alternatively, instead of adding ReadGuard::upgradeToWrite(), maybe add a ctor to WriteGuard taking ReadGuard, that will steal the mutex from the ReadGuard and upgrade it to write?

Reply:
I do actually like the semantics of Denis' proposal here; WriteGuard(ReadGuard&, TryToLock{}) might be a bit awkward to implement but the semantics feel a lot better than a ReadGuard that owns a write lock.
We definitely want to have a way to go straight to write lock via guards for users who know they're going to need the write lock and don't want to pay the price of read lock + upgrade.

Reply:
@demin80, yes however the ReadGuard vs WriteGuard indicates intent more clearly. If your code only needs to write (i.e. a producer) then you can only use a WriteGuard. With the WriteGuard taking a ReadGuard and upgrading I feel you have to create 2 guard objects instead of one.

Reply:
I personally prefer either the two distinct guards (with the reader having an upgrade) or a single guard which can upgrade as needed.